File: src/main/java/list/KthNode.java
package list;

import structure.ListNode;

/**
 * https://leetcode-cn.com/problems/kth-node-from-end-of-list-lcci/
 */
public class KthNode {
    public static void main(String[] args) {
        ListNode list = new ListNode(1);
        ListNode root = list;
        for (int i = 2; i < 6; i++) {
            list.next = new ListNode(i);
            list = list.next;
        }

        System.out.println(new KthNode().kthToLast(root, 3));
    }

    public int kthToLast(ListNode head, int k) {
        ListNode right = head;
        ListNode left = head;

        for (int i = 1; i < k; i++) {
            right = right.next;
        }

        while (right.next != null){
            right = right.next;
            left = left.next;
        }

        return left.val;
    }
}

File: CppAlgorithms/list/KthNode.cpp
#include <vector>
#include <iostream>
#include "../structure/ListNode.h"
using namespace std;

/**
 * https://leetcode-cn.com/problems/kth-node-from-end-of-list-lcci/
 * 实现一种算法，找出单向链表中倒数第 k 个节点。返回该节点的值。
 * 输入： 1->2->3->4->5 和 k = 2
 * 输出： 4
 */

class Solution {
public:
    int kthToLast(ListNode* head, int k) {
        ListNode *first = head;
        ListNode *last = head;
        for (int i = 0; i < k; ++i) {
            last = last->next;
        }
        while (last){
            last = last->next;
            first = first->next;
        }
        return first->val;
    }
};


int main() {
    Solution s;
    vector<int> v{1, 2, 3, 4, 5};
    ListNode listNode(v);
    cout << s.kthToLast(&listNode, 2) << endl;
}
